Add Adjust text size to demo and fix demo to support new abstraction layer

thesocialdev opened the pull request.

mateusbs17 commented:
@berndsi this patch introduces some fixes to the All transformation demo and add Adjust Text Size control.

berndsi commented on demo/lib/TransformationControls/addFooterControl.js:
```diff
@@ -101,9 +101,13 @@ export default [
-    addFooterReadMore(iframeWindow, iframeDocument, iframe)
-    addFooterLegal(iframeWindow, iframeDocument)
+    if (iframeWindow.pagelib.c1) {
+      // TODO: footer transformations are not exposed in the abstraction layer
```
correct, coming in #208

mateusbs17 replied:
Probably better to wait for the PR to be merged then.

berndsi commented on demo/lib/TransformationControls/addLeadIntroductionControl.js:
```diff
@@ -2,18 +2,24 @@
+  // eslint-disable-next-line no-undef
+  if (iframeDocument.articleRef.sourceType === ArticleRefSourceType.mobileView) {
+    if (iframeWindow.pagelib.c1) {
+      // TODO: moveLeadIntroductionUp is not exposed in the abstraction layer
```
We're not going to add this AFAIK. The service side already does this, so there is no reason to expose it.

berndsi commented on demo/lib/TransformationControls/addHideRedlinksControl.js:
```diff
@@ -1,12 +1,18 @@
+const hideRedLinks = (iframeWindow, iframeDocument) => {
+  event.target.disabled = true
+  if (iframeWindow.pagelib.c1) {
+    // TODO: hideRedLinks is not exposed in the abstraction layer
```
We're not going to add this AFAIK. The service side already does this, so there is no reason to expose it.

berndsi commented on demo/lib/TransformationControls/addTableCollapseControl.js:
```diff
@@ -1,14 +1,19 @@
+const collapseTables = (iframeWindow, iframeDocument) => {
+  event.target.disabled = true
+  if (iframeWindow.pagelib.c1) {
+    // TODO: collapseTables is not exposed in the abstraction layer
```
FYI: It's available in the abstraction layer only as part of the `setMulti` call. The apps said they wouldn't need to change the value after the the page is loaded. That's why there is no separate function. Not sure if that will change in the future but likely won't be happening soon.

mateusbs17 commented on demo/DemoArticles/data/de.Schweden.167705875.PCS.html:
```diff
@@ -2078,8 +2078,12 @@
-lazyLoader.loadPlaceholders();
-pagelib.CollapseTable.setupEventHandling(window, document, false, null);
+if (pagelib.c1) {
+    pagelib.c1.PageMods.onPageLoad(window,document);
```
You're right. I didn't test it added the onPageLoad event.

berndsi replied:
Did you make those changes manually to the files? Those files are generated and will be overridden when someone calls the `update.js` script.

mateusbs17 replied:
Good catch, I ran the script and the updated versions are working fine.

berndsi commented:
The Text Size Adjustment implementation seems to be only for for certain browsers. Would you document what it takes for it to work? Also needs mention in pcs.md.

thesocialdev commented on demo/lib/TransformationControls/addFooterControl.js:
```diff
@@ -100,9 +100,39 @@ export default [
+    if (iframeWindow.pagelib.c1) {
+      iframeWindow.pagelib.c1.Footer.add(
+        document,
+        'Knight Lore',  // TODO: expose the article for each closure
```
@berndsi that's the only thing I wasn't able to do in this patch. Do you have any ideas on how to handle it?

berndsi replied:
Which part? adding the footer or getting the title?

thesocialdev replied:
Getting the title. The footer is added, although its UI is different from MCS.

thesocialdev commented:
> The Text Size Adjustment implementation seems to be only for for certain browsers. Would you document what it takes for it to work? Also needs mention in pcs.md.

Done.

berndsi commented:
Thanks for the update. I still have a hard time getting this to work.
(In Safari the whole AllTransforms.html page doesn't even display properly. That's probably out of scope for this PR.)
In Chrome on Mac I see that it does add `style="text-size-adjust: 20%;"` if I use `20%` but I see no change in the font size appearance.
In Firefox Developer Edition on Mac it doesn't even set that style.

thesocialdev commented:
> Thanks for the update. I still have a hard time getting this to work.
> (In Safari the whole AllTransforms.html page doesn't even display properly. That's probably out of scope for this PR.)
> In Chrome on Mac I see that it does add `style="text-size-adjust: 20%;"` if I use `20%` but I see no change in the font size appearance.

I observed the same behavior, I assume that this can't be tested in the AllTransforms demo. The description at MSDN highlight that this is used for smartphones and tablets:
>The text-size-adjust CSS property controls the text inflation algorithm used on **some smartphones and tablets**. Other browsers will ignore this property.

Maybe this PR is not valid then?

> In Firefox Developer Edition on Mac it doesn't even set that style.

It looks that Firefox doesn't support percentage values for text-size-adjust [1]

[1] https://developer.mozilla.org/en-US/docs/Web/CSS/text-size-adjust#Browser_compatibility

berndsi commented:
> Maybe this PR is not valid then?

I think the "fix demo to support new abstraction layer" part is still useful regardless. It would be easier for reviewers if the update of the page data was done separately. I don't remember why it was needed for this PR.

berndsi commented:
I also want to get the changes in the pcs.md doc merged.

thesocialdev commented:
> > Maybe this PR is not valid then?
> 
> [...] It would be easier for reviewers if the update of the page data was done separately. [...]

Done.

berndsi commented on demo/lib/TransformationControls/addWidenImagesControl.js:
```diff
@@ -8,6 +8,13 @@ const shouldWidenImage = image =>
 
+const widenImage = (iframeWindow, image) => {
+  if (iframeWindow.pagelib.c1) {
+    // TODO: widen image transform is not exposed in the abstraction layer
```
You could remove the TODO. widen image transform is already done server-side.

berndsi commented on demo/lib/TransformationControls/addTextSizeAdjustmentPercentage.js:
```diff
@@ -0,0 +1,17 @@
+
+export default [
+  'adjust_text_size_form',
+  'Text Size Adjustment (Percentage) *',
```
maybe add a comment to say that this only works on certain mobile browsers with the caniuse link?

thesocialdev replied:
You mean a comment in the code or in the sidebar to be showed in the demo interface?

berndsi replied:
Sorry. The latter. I meant to add some user visible text to this input form field. So that it's clearer to demo users why this doesn't work.

thesocialdev replied:
done.

berndsi commented:
Looks good to me. 

There's a conflicting PR #227 which changes PageMods to Page and removes the `document` parameter for the functions. Would it be ok for you to rebase your changes on top of that or should we merge this one first and adjust the other PR?

thesocialdev commented:
> Would it be ok for you to rebase your changes on top of that or should we merge this one first and adjust the other PR?

It's better solving conflicts in this PR, removing the document is already going to break some functionalities in the demo.

berndsi commented:
Sounds good. I just merged the other PR. Sorry for making you rebase again.

thesocialdev commented:
> Sounds good. I just merged the other PR. Sorry for making you rebase again.

Cool, I rebased and updated the PageMods part from my PR. But can't tell if the change from #227 will work in the AllTransforms demo until it's deployed.

berndsi commented:
Good point. We should look into making the demo use the local versions of the pagelib CSS and JS.
